File: Example 14 JNA/src/Example14.java
// Projeto : JNA Example #14
// Classe: Example14.java
// Função : : Receive a Struct Containing an Array of Strings from C
// Autor(es) : Hugo Dionizio Santos
// Data : Sáb Mai 16 16:27:44 BRT 2015

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.ptr.PointerByReference;		
import com.sun.jna.Pointer;
import com.sun.jna.Structure;
import com.sun.jna.Structure.ByReference;

public class Example14 {
	public interface CLibrary extends Library {
		public static class Example14Struct extends Structure {
			public static class ByValue extends Example14Struct implements Structure.ByValue {}
		
			public int numVals;
			public Pointer vals; // char**
		}

		public Example14Struct.ByValue example14_getStrings();
		public void example14_cleanup(Example14Struct.ByValue sVal);
}

	static public void main(String argv[]) {
		CLibrary clib;
		CLibrary.Example14Struct.ByValue ex14val;

		try {
			clib = (CLibrary)Native.loadLibrary("./libexample14.so", CLibrary.class);
			// Receive a Struct Containing an Array of Strings from C
			ex14val = clib.example14_getStrings();
		}
		catch (UnsatisfiedLinkError e) {
			clib = (CLibrary)Native.loadLibrary("./bin/libexample14.so", CLibrary.class);
			// Receive a Struct Containing an Array of Strings from C
			ex14val = clib.example14_getStrings();
		}

		System.out.println("example 14: retrieved " + ex14val.numVals + " values:");

		// getStringArray copies the contents of the C-allocated memory buffer into a Java-managed String[]
		String[] ex14vals = ex14val.vals.getStringArray(0, ex14val.numVals);
		for (int ex14loop=0; ex14loop<ex14val.numVals; ex14loop++) {
			System.out.println("\t" + ex14vals[ex14loop]);
		}
		System.out.println("\t(example 14 cleanup)");
		clib.example14_cleanup(ex14val);
	}
}

// Fim da classe

File: Example 14 JNA/src/example14.c
// Projeto : JNA Example #14
// Algoritmo: example14.c
// Função : : Receive a Struct Containing an Array of Strings from C
// Autor(es) : Hugo Dionizio Santos (adaptações e correções), Ethan Shayne in http://www.eshayne.com/jnaex
// Data : Sáb Mai 16 16:27:44 BRT 2015

#include <stdlib.h>
#include <string.h>

// Seção de Variáveis Globais

// Interface do algoritmo

// Função Principal
//	Seção de Variáveis Locais
//	Seção de Comandos

// Seção de Funções
typedef struct Example14Struct {
	int numVals;
	char** vals;
} Example14Struct;

Example14Struct example14_getStrings()
{
	Example14Struct sVal;
	sVal.numVals = 3;
	sVal.vals = (char**)malloc(sizeof(char*) * sVal.numVals);
	memset(sVal.vals, 0, sizeof(char*) * sVal.numVals);

	sVal.vals[0] = (char*)malloc(sizeof(char) * 4);
	memset(sVal.vals[0], 0, sizeof(char) * 4);
	strcpy(sVal.vals[0], "ten");

	sVal.vals[1] = (char*)malloc(sizeof(char) * 7);
	memset(sVal.vals[1], 0, sizeof(char) * 7);
	strcpy(sVal.vals[1], "eleven");

	sVal.vals[2] = (char*)malloc(sizeof(char) * 7);
	memset(sVal.vals[2], 0, sizeof(char) * 7);
	strcpy(sVal.vals[2], "twelve");

	return sVal;
}

void example14_cleanup(Example14Struct sVal)
{
	int loop = 0;
	for (loop=0; loop>sVal.numVals; loop++)
	{
		free(sVal.vals[loop]);
	}
	free(sVal.vals);
}
// Fim do algoritmo
